cirq/ops/projectors.py

```diff
@@ -0,0 +1,75 @@
+from typing import Any, Dict, Iterable, List, Tuple, TYPE_CHECKING, Union
+
+import numpy as np
+
+from cirq import value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class Projector(raw_types.Gate):
+    """A non-unitary gate that projects onto one or more qubits."""
+
+    def __init__(self,
+                 projection_basis: Union[List[List[float]], np.ndarray],
+                 qid_shape: Tuple[int, ...] = (2,)):
+        """
+        Args:
+            projection_basis: a (2**num_qubits, p) matrix that lists the
+                projection vectors.
+            qid_shape: Specifies the dimension of the projection. The default is
+                2 (qubit).
+
+        Raises:
+            ValueError: If the basis vector is empty.
+        """
+        projection_array = np.asarray(projection_basis)
+
+        if len(projection_array.shape) != 2:
+            raise ValueError('The input projection_basis must be a 2D array')
+
+        if np.prod(qid_shape) != projection_array.shape[1]:
```

Comment:
Don't I already have the test in there?
`
test_projector_overcomplete_basis
`

It throws an LinAlgError.

Reply:
Indeed.

Reply:
Actually, it'd be better to check whether projection basis is linearly independent in the constructor rather than wait until an error is triggered in `_channel_`.